csrc/scheduler/reduction_utils.cpp

```diff
@@ -30,9 +31,11 @@ TensorView* scheduleReductionTV(
   // Inner here though is only relative to the other axis. When
   // rparams.fastest_dim == false, the reduction axis is logically outside the
   // iteration axis.
-  const int iter_axis = 0;
+  const int has_outmost_dim_sharded = isSharded(reduction_tv);
+  const int iter_axis = has_outmost_dim_sharded ? 1 : 0;
   const int outer_reduce_axis = rparams.schedule_3D ? 1 : 0;
-  const int inner_reduce_axis = rparams.schedule_3D ? 2 : has_iter_axis ? 1 : 0;
+  const int inner_reduce_axis =
+      rparams.schedule_3D ? 2 : has_outmost_dim_sharded + has_iter_axis;
```

Comment:
When `rparams.schedule_3D` is true, shouldn't this be 3 when sharded?

Reply:
we assert in the code that we don't mix 3D reduction and multidevice. I mainly assume this for simplicity. Is it a problem?
Btw: why is 3D important in general?

Reply:
I add an assert at this place too